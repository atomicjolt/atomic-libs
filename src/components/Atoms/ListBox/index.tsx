import React, { useMemo, useRef, useState } from "react";
import type { AriaListBoxProps } from "react-aria";
import { useListState, Node, ListState } from "react-stately";
import {
  mergeProps,
  useFilter,
  useFocusRing,
  useListBox,
  useListBoxSection,
  useOption,
} from "react-aria";
import {
  List,
  ListItem,
  SectionSeperator,
  SectionTitle,
  SubList,
} from "./ListBox.styles";
import Label from "../Label";
import { BaseProps } from "../../../types";
import classNames from "classnames";
import useForwardedRef from "../../../hooks/useForwardedRef";
import SearchInput from "../../Inputs/SearchInput";

export type ListBoxProps<T> = AriaListBoxProps<T> &
  BaseProps & {
    /** Allow the items in the select to be searchable */
    isSearchable?: boolean;
    /** Placeholder for the search input */
    searchPlaceholder?: string;
  };

/** A listbox displays a list of options and allows a user to select one or more of them.
 * Used as the dropdown menu for `ComboBox` and `CustomSelect` */
export function ListBox<T extends object>(props: ListBoxProps<T>) {
  const state = useListState(props);
  return <UnmanagedListBox {...props} state={state} />;
}

export type UnmanagedListBoxProps<T> = ListBoxProps<T> & {
  state: ListState<T>;
};

/** Listbox, but the state is passed in instead of managed internally */
export const UnmanagedListBox = React.forwardRef<
  HTMLUListElement,
  UnmanagedListBoxProps<any>
>((props, ref) => {
  const {
    isSearchable,
    searchPlaceholder,
    state,
    className,
    size = "medium",
  } = props;
  const internalRef = useForwardedRef(ref);
  const { listBoxProps, labelProps } = useListBox(props, state, internalRef);
  const [searchValue, setSearchValue] = useState("");

  const { contains } = useFilter({ sensitivity: "base" });

  const filteredItems = useMemo(
    () =>
      [...state.collection].filter((item) =>
        item.type === "section" ? true : contains(item.textValue, searchValue)
      ),
    [state.collection, searchValue]
  );

  return (
    <>
      <Label {...labelProps}>{props.label}</Label>
      <List
        {...listBoxProps}
        ref={internalRef}
        className={classNames("aje-listbox", className, `is-${size}`)}
      >
        {isSearchable && (
          <SearchInput
            label="Search"
            hideLabel
            value={searchValue}
            onChange={setSearchValue}
            placeholder={searchPlaceholder}
            size={size}
          />
        )}
        {filteredItems.map((item) =>
          item.type === "section" ? (
            <ListBoxSection
              key={item.key}
              section={item}
              state={state}
              filter={(v) => contains(v, searchValue)}
            />
          ) : (
            <ListBoxOption key={item.key} item={item} state={state} />
          )
        )}
      </List>
    </>
  );
});

interface ListBoxSectionProps<T> {
  readonly section: Node<T>;
  readonly state: ListState<T>;
  readonly filter: (text: string) => boolean;
}

function ListBoxSection<T>(props: ListBoxSectionProps<T>) {
  const { section, state, filter } = props;
  const { itemProps, headingProps, groupProps } = useListBoxSection({
    heading: section.rendered,
    "aria-label": section["aria-label"],
  });

  const children = [...section.childNodes].filter((item) =>
    filter(item.textValue)
  );

  if (children.length === 0) {
    return null;
  }

  // If the section is not the first, add a separator element to provide visual separation.
  // The heading is rendered inside an <li> element, which contains
  // a <ul> with the child items.
  return (
    <>
      {section.key !== state.collection.getFirstKey() && (
        <SectionSeperator role="presentation" />
      )}
      <li {...itemProps}>
        {section.rendered && (
          <SectionTitle {...headingProps}>{section.rendered}</SectionTitle>
        )}
        <SubList {...groupProps}>
          {children.map((node) => (
            <ListBoxOption key={node.key} item={node} state={state} />
          ))}
        </SubList>
      </li>
    </>
  );
}

interface ListBoxOptionProps<T> {
  item: Node<T>;
  state: ListState<T>;
}

function ListBoxOption<T>(props: ListBoxOptionProps<T>) {
  const { item, state } = props;
  const ref = useRef(null);
  const { optionProps } = useOption({ key: item.key }, state, ref);

  // Determine whether we should show a keyboard
  // focus ring for accessibility
  const { focusProps } = useFocusRing();

  return (
    <ListItem
      {...mergeProps(optionProps, focusProps)}
      ref={ref}
      data-focus-visible={item.key === state.selectionManager.focusedKey}
    >
      {item.rendered}
    </ListItem>
  );
}
