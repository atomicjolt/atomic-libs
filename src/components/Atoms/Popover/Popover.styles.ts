import styled from "styled-components";

export const PopoverWrapper = styled.div`
  position: relative;
`;

export const PopoverUnderlay = styled.div`
  position: fixed;
  inset: 0;
`;

export const PopoverContent = styled.div``;
