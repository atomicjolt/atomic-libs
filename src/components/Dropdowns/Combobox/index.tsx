import React, { useState, useEffect } from 'react'
import '../../general.scss'
import './styles.scss'

export interface Props {
  children: React.ReactNode;
  /** Must include a label. Labels are always Sentence case. */
  label: string;
  /** Error text should be descriptive and explicit in meaning. */
  error?: string;
  /** For additional information (ex. date format mm/dd/yy) */
  message?: string;
  /** Only use in very specific circumstances. This hides the label from view, but still allows screen readers to read the label. (A filter dropdown with a clear meaning could potentially be a use case.) */
  hideLabel?: boolean;
  /** The select size should reflect the size of its content. */
  size?: 'small' | 'medium' | 'large' | 'full' | 'auto';
  disabled?: boolean;
  required?: boolean;
}

/** 
 * Combobox
 * 
 * https://www.w3.org/TR/wai-aria-practices/examples/combobox/combobox-select-only.html for accessibility implementation.
 * */
export default function Combobox({ 
    children,
    size = 'medium', 
    label, 
    error, 
    message, 
    hideLabel = false,
    disabled = false,
  }: Props) {

  const errorID = 'errorText';
  /* Add a space before the added class rather than inside the className attr on the tag. Looks cleaner. */
  let errorClass = error ? ' has-error' : '';
  let disabledClass = disabled ? ' is-disabled' : '';
  let hiddenClass = hideLabel ? ' aj-hidden' : '';

  const [menuActive, setMenuActive] = useState(false)

  const handleOpen = () => {
    if (menuActive) {
      setMenuActive(false)
    } else {
      setMenuActive(true)
    }
  }

  useEffect(() => {
    const closeMenu = () => {
      if (!menuActive) return
      setMenuActive(false)
    }

    window.addEventListener("click", closeMenu)

    return () => window.removeEventListener("click", closeMenu)
  }, [menuActive])

  return (
    <div className={`aj-dropdown is-${size}${errorClass}${disabledClass}`}>
      <label className={`aj-label${hiddenClass}`} id="comboLabel">
        {label}
        {message ?
          <p className="aj-label--message">{message}</p>
          : null
        }
      </label>
      <div className="aj-combobox">
        <div 
          className="aj-combobox__input is-searchable"
          aria-owns="listboxID"
          aria-expanded={menuActive}
          aria-haspopup="listbox"
          id="comboID"
          role="combobox"
          onClick={handleOpen}
        >
          <input
            type="text"
            aria-autocomplete='both'
            aria-controls='listboxID'
            aria-labelledby='comboLabel'
            aria-describedby={error ? errorID : ''}
            id="comboInputID"
          />
        </div>
        <ul
          className="aj-combobox__menu"
          role="listbox"
          id="lisboxID"
          aria-labelledby="comboLabel"
        >
          <li
            className="aj-combobox__option is-focused"
            role="option"
            id="op1"
          >
            Option
          </li>
          <li
            className="aj-combobox__option"
            role="option"
            id="op2"
          >
            {children}
          </li>
          <li
            className="aj-combobox__option"
            role="option"
            id="op3"
          >
            {children}
          </li>
        </ul>
      </div>
      {error ?
        <p id={errorID} className="aj-label--error">{error}</p>
        : null
      }
    </div>
  )
}
