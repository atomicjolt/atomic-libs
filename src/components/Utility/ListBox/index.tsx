import React, { useRef } from "react";
import type { AriaListBoxProps } from "react-aria";
import { useListState, Node, ListState } from "react-stately";
import {
  mergeProps,
  useFocusRing,
  useListBox,
  useListBoxSection,
  useOption,
} from "react-aria";
import {
  List,
  ListItem,
  SectionSeperator,
  SectionTitle,
  SubList,
} from "./ListBox.styles";
import Label from "../Label";
import { BaseProps } from "../../../types";
import classNames from "classnames";
import useForwardedRef from "../../../hooks/useForwardedRef";

export type ListBoxProps<T> = AriaListBoxProps<T> & BaseProps;

/** A listbox displays a list of options and allows a user to select one or more of them.
 * Used as the dropdown menu for `ComboBox` and `CustomSelect` */
export function ListBox<T extends object>(props: ListBoxProps<T>) {
  const state = useListState(props);
  return <UnmanagedListBox {...props} state={state} />;
}

export type UnmanagedListBoxProps<T> = ListBoxProps<T> & {
  state: ListState<T>;
};

/** Listbox, but the state is passed in instead of managed internally */
export const UnmanagedListBox = React.forwardRef<
  HTMLUListElement,
  UnmanagedListBoxProps<any>
>((props, ref) => {
  const { state, className, size = "medium" } = props;
  const internalRef = useForwardedRef(ref);
  const { listBoxProps, labelProps } = useListBox(props, state, internalRef);

  return (
    <>
      <Label {...labelProps}>{props.label}</Label>
      <List
        {...listBoxProps}
        ref={internalRef}
        className={classNames("aje-listbox", className, `is-${size}`)}
      >
        <input role="searchbox" />
        {[...state.collection].map((item) =>
          item.type === "section" ? (
            <ListBoxSection key={item.key} section={item} state={state} />
          ) : (
            <ListBoxOption key={item.key} item={item} state={state} />
          )
        )}
      </List>
    </>
  );
});

interface ListBoxSectionProps<T> {
  section: Node<T>;
  state: ListState<T>;
}

function ListBoxSection<T>(props: ListBoxSectionProps<T>) {
  const { section, state } = props;
  let { itemProps, headingProps, groupProps } = useListBoxSection({
    heading: section.rendered,
    "aria-label": section["aria-label"],
  });

  // If the section is not the first, add a separator element to provide visual separation.
  // The heading is rendered inside an <li> element, which contains
  // a <ul> with the child items.
  return (
    <>
      {section.key !== state.collection.getFirstKey() && (
        <SectionSeperator role="presentation" />
      )}
      <li {...itemProps}>
        {section.rendered && (
          <SectionTitle {...headingProps}>{section.rendered}</SectionTitle>
        )}
        <SubList {...groupProps}>
          {[...section.childNodes].map((node) => (
            <ListBoxOption key={node.key} item={node} state={state} />
          ))}
        </SubList>
      </li>
    </>
  );
}

interface ListBoxOptionProps<T> {
  item: Node<T>;
  state: ListState<T>;
}

function ListBoxOption<T>(props: ListBoxOptionProps<T>) {
  const { item, state } = props;
  let ref = useRef(null);
  let { optionProps } = useOption({ key: item.key }, state, ref);

  // Determine whether we should show a keyboard
  // focus ring for accessibility
  let { focusProps } = useFocusRing();

  return (
    <ListItem
      {...mergeProps(optionProps, focusProps)}
      ref={ref}
      data-focus-visible={item.key === state.selectionManager.focusedKey}
    >
      {item.rendered}
    </ListItem>
  );
}
