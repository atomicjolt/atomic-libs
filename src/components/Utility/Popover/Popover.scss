@use "../../mixins.scss";

.aje-popover__wrapper {
  position: relative;
  display: inline-block;
}

.aje-popover {
  --popover-animation: transform 200ms ease, opacity 200ms;
  --popover-disatnce: 5px;

  visibility: hidden;
  position: absolute;
  width: var(--popover-width, auto);
  z-index: 100;
  opacity: 0;
  transform: scaleY(0.9);
  transform-origin: top;
  transition: var(--popover-animation);

  &-bottom-left {
    margin-top: var(--popover-disatnce);
    top: 100%;
    left: 0;
  }

  &-bottom-right {
    margin-top: var(--popover-disatnce);
    top: 100%;
    right: 0;
  }

  &-top-left {
    margin-bottom: var(--popover-disatnce);
    bottom: 100%;
    left: 0;
  }

  &-top-right {
    margin-bottom: var(--popover-disatnce);
    bottom: 100%;
    right: 0;
  }

  &-left {
    margin-right: var(--popover-disatnce);
    right: 100%;
    top: 0;
  }

  &-right {
    margin-left: var(--popover-disatnce);
    left: 100%;
    top: 0;
  }

  &.is-small {
    --popover-width: 100px;
  }
  &.is-medium {
    --popover-width: 200px;
  }
  &.is-large {
    --popover-width: 300px;
  }
  &.is-full {
    --popover-width: 100%;
  }
  &.is-auto {
    --popover-width: max-content;
  }

  &.is-visible {
    visibility: visible;
    opacity: 1;
    transform: scaleY(1);
  }

  .dropdown {
    background-color: lightblue;
    padding: 10px;
    @include mixins.regular;
    font-size: 20px;
  }
}
