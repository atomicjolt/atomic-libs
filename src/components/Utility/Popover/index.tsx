import React from "react";
import cn from "classnames";
import { Sizes } from "../../../types";

export type PopoverPosition =
  | "bottom-right"
  | "bottom-left"
  | "top-right"
  | "top-left"
  | "left"
  | "right";

export interface PopoverProps {
  children: React.ReactNode;
  size?: Sizes;
  show?: boolean;
  position?: PopoverPosition;
}

export default function Popover(props: PopoverProps) {
  const { children, show, size, position = "bottom-left" } = props;

  return (
    <div
      className={cn("aje-popover", `is-${size}`, `aje-popover-${position}`, {
        "is-visible": show,
      })}
    >
      {children}
    </div>
  );
}

interface PopoverWrapperProps {
  children: React.ReactNode;
}

export function PopoverWrapper(props: PopoverWrapperProps) {
  const { children } = props;

  return <div className={cn("aje-popover__wrapper")}>{children}</div>;
}
