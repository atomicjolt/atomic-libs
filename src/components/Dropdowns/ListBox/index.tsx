import React, { useMemo, useRef, useState } from "react";
import type { AriaListBoxProps } from "react-aria";
import { useListState, Node, ListState } from "react-stately";
import {
  mergeProps,
  useFilter,
  useFocusRing,
  useListBox,
  useListBoxSection,
  useOption,
} from "react-aria";
import classNames from "classnames";
import { BaseProps } from "../../../types";
import useForwardedRef from "../../../hooks/useForwardedRef";
import SearchInput from "../../Inputs/SearchInput";
import MaterialIcon from "../../Icons/MaterialIcon";
import { Label } from "../../Fields";
import { Divider } from "@components/Layout/Divider";
import { List, ListItem, SectionTitle, SubList } from "./ListBox.styles";

export type ListBoxProps<T> = AriaListBoxProps<T> &
  BaseProps & {
    /** Allow the items in the select to be searchable */
    isSearchable?: boolean;
    /** Placeholder for the search input */
    searchPlaceholder?: string;
    /** Whether to show a checkmark next to selected items */
    showCheckmark?: boolean;
  };

/** A listbox displays a list of options and allows a user to select one or more of them.
 * Used as the dropdown menu for `ComboBox` and `CustomSelect` */
export function ListBox<T extends object>(props: ListBoxProps<T>) {
  const state = useListState(props);
  return <UnmanagedListBox {...props} state={state} />;
}

export type UnmanagedListBoxProps<T> = Omit<ListBoxProps<T>, "children"> & {
  state: ListState<T>;
};

/** Listbox, but the state is passed in instead of managed internally */
export const UnmanagedListBox = React.forwardRef<
  HTMLUListElement,
  UnmanagedListBoxProps<any>
>((props, ref) => {
  const {
    isSearchable,
    searchPlaceholder = "Search",
    state,
    className,
    size = "medium",
    showCheckmark = false,
  } = props;
  const internalRef = useForwardedRef(ref);
  const { listBoxProps, labelProps } = useListBox(props, state, internalRef);
  const [searchValue, setSearchValue] = useState("");
  const [isSearching, setIsSearching] = useState(false);

  const { contains } = useFilter({ sensitivity: "base" });

  const filteredItems = useMemo(
    () =>
      [...state.collection].filter((item) =>
        item.type === "section" ? true : contains(item.textValue, searchValue)
      ),
    [state.collection, searchValue]
  );

  if (isSearching) {
    // Prevent the listbox from handling keyboard events while the search input is focused
    // Stops it from taking focus away from the search input
    delete listBoxProps.onKeyDownCapture;
  }

  return (
    <>
      <Label {...labelProps}>{props.label}</Label>
      <List
        {...listBoxProps}
        ref={internalRef}
        className={classNames("aje-listbox", className, `is-${size}`)}
      >
        {isSearchable && (
          <SearchInput
            aria-label="Search"
            value={searchValue}
            onChange={setSearchValue}
            placeholder={searchPlaceholder}
            size={size}
            onFocus={() => setIsSearching(true)}
            onBlur={() => setIsSearching(false)}
          />
        )}
        {filteredItems.map((item) =>
          item.type === "section" ? (
            <ListBoxSection
              key={item.key}
              section={item}
              state={state}
              filter={(v) => contains(v, searchValue)}
              showCheckmark={showCheckmark}
            />
          ) : (
            <ListBoxOption
              key={item.key}
              item={item}
              state={state}
              showCheckmark={showCheckmark}
            />
          )
        )}
      </List>
    </>
  );
});

interface ListBoxSectionProps<T> {
  readonly section: Node<T>;
  readonly state: ListState<T>;
  readonly filter: (text: string) => boolean;
  readonly showCheckmark?: boolean;
}

function ListBoxSection<T>(props: ListBoxSectionProps<T>) {
  const { section, state, filter, showCheckmark = false } = props;
  const { itemProps, headingProps, groupProps } = useListBoxSection({
    heading: section.rendered,
    "aria-label": section["aria-label"],
  });

  const children = [...section.childNodes].filter((item) =>
    filter(item.textValue)
  );

  if (children.length === 0) {
    return null;
  }

  // If the section is not the first, add a separator element to provide visual separation.
  // The heading is rendered inside an <li> element, which contains
  // a <ul> with the child items.
  return (
    <>
      {section.key !== state.collection.getFirstKey() && <Divider as="li" />}
      <li {...itemProps}>
        {section.rendered && (
          <SectionTitle {...headingProps}>{section.rendered}</SectionTitle>
        )}
        <SubList {...groupProps}>
          {children.map((node) => (
            <ListBoxOption
              key={node.key}
              item={node}
              state={state}
              showCheckmark={showCheckmark}
            />
          ))}
        </SubList>
      </li>
    </>
  );
}

interface ListBoxOptionProps<T> {
  item: Node<T>;
  state: ListState<T>;
  showCheckmark?: boolean;
}

function ListBoxOption<T>(props: ListBoxOptionProps<T>) {
  const { item, state, showCheckmark } = props;
  const ref = useRef(null);
  const { optionProps } = useOption({ key: item.key }, state, ref);

  // Determine whether we should show a keyboard
  // focus ring for accessibility
  const { focusProps } = useFocusRing();

  return (
    <ListItem
      {...mergeProps(optionProps, focusProps)}
      ref={ref}
      data-focus-visible={item.key === state.selectionManager.focusedKey}
    >
      {item.rendered}
      {showCheckmark && <MaterialIcon icon="check" />}
    </ListItem>
  );
}
