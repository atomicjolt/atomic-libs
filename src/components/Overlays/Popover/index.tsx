import React, { useCallback, useContext, useMemo, useState } from "react";
import {
  AriaPopoverProps,
  DismissButton,
  Overlay,
  usePopover,
} from "@react-aria/overlays";
import { OverlayTriggerProps, useOverlayTriggerState } from "react-stately";
import { mergeProps } from "@react-aria/utils";

import { HasClassName, HasVariant } from "../../../types";
import { useRenderProps } from "@hooks";
import { useContextProps } from "@hooks/useContextProps";
import { useResizeObserver } from "@hooks/useResizeObserver";
import { useForwardedRef } from "@hooks/useForwardedRef";
import { PopoverUnderlay, PopoverContent } from "./Popover.styles";
import { PopoverContext } from "./context";
import { OverlayTriggerStateContext } from "../OverlayTrigger/context";
import { invertPlacementAxis } from "@utils/placement";

export interface PopoverProps
  extends Omit<AriaPopoverProps, "popoverRef" | "triggerRef">,
    OverlayTriggerProps,
    HasClassName,
    HasVariant<"listbox" | "menu" | "datepicker"> {
  children: React.ReactNode;
  /** A ref to the element that triggers the popover to appear. Not necessary when used with trigger wrapper */
  triggerRef?: React.RefObject<HTMLElement>;
  id?: string;
}

/** A popover is an overlay element positioned relative to a target. */
export const Popover = React.forwardRef<HTMLDivElement, PopoverProps>(
  (props: PopoverProps, ref) => {
    const contextProps = useContextProps(PopoverContext, props);
    const {
      children,
      offset = 8,
      triggerRef,
      variant,
      placement = "bottom",
      ...rest
    } = contextProps;
    const contextState = useContext(OverlayTriggerStateContext);
    const localState = useOverlayTriggerState(props);
    const state = contextState ?? localState;

    const internalRef = useForwardedRef(ref);

    const { popoverProps, underlayProps } = usePopover(
      {
        ...rest,
        placement,
        offset,
        triggerRef: triggerRef!,
        popoverRef: internalRef,
      },
      state
    );

    const [triggerWidth, setTriggerWidth] = useState(
      `${triggerRef?.current?.offsetWidth || 0}px`
    );

    useResizeObserver(
      triggerRef?.current,
      useCallback(
        () => setTriggerWidth(`${triggerRef?.current?.offsetWidth || 0}px`),
        []
      )
    );

    const renderProps = useRenderProps({
      componentClassName: "aje-popover",
      ...contextProps,
    });

    const transformOrigin = invertPlacementAxis(placement);

    if (!state.isOpen) {
      return null;
    }

    return (
      <Overlay>
        <PopoverUnderlay {...underlayProps} className="aje-popover-underlay" />
        <PopoverContent
          ref={internalRef}
          id={rest.id}
          {...mergeProps(popoverProps, renderProps)}
          style={{
            ...popoverProps.style,
            ...renderProps.style,
            // @ts-ignore
            "--trigger-width": triggerWidth,
          }}
          $transformOrigin={transformOrigin}
        >
          <DismissButton onDismiss={state.close} />
          {children}
          <DismissButton onDismiss={state.close} />
        </PopoverContent>
      </Overlay>
    );
  }
);
