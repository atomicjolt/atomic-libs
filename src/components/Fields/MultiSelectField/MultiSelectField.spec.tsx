import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import { Popover } from "@components/Overlays/Popover";
import { ListBox } from "@components/Dropdowns/ListBox";
import { MultiSelectField } from ".";

describe("MultiSelectField", () => {
  describe("matches snapshots", () => {
    test("when closed", () => {
      const result = render(
        <MultiSelectField>
          <MultiSelectField.Button>Open up</MultiSelectField.Button>
          <Popover>
            <ListBox>
              <ListBox.Item>Item 1</ListBox.Item>
              <ListBox.Item>Item 2</ListBox.Item>
              <ListBox.Item>Item 3</ListBox.Item>
            </ListBox>
          </Popover>
        </MultiSelectField>
      );
      expect(result.asFragment()).toMatchSnapshot();
    });

    test("when open", () => {
      const result = render(
        <MultiSelectField isOpen>
          <MultiSelectField.Button>Open up</MultiSelectField.Button>
          <Popover>
            <ListBox>
              <ListBox.Item>Item 1</ListBox.Item>
              <ListBox.Item>Item 2</ListBox.Item>
              <ListBox.Item>Item 3</ListBox.Item>
            </ListBox>
          </Popover>
        </MultiSelectField>
      );
      expect(result.asFragment()).toMatchSnapshot();
    });
  });
});
