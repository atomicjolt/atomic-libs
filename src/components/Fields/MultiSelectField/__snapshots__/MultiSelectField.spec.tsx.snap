// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiSelectField > matches snapshots > when closed 1`] = `
<DocumentFragment>
  <div
    class="sc-hzhKNl gQkKli aje-multiselect"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="react-aria-:r7: react-aria-:r3:"
      class="sc-dLNtp sc-feUYzb hkAcSC fzpzpe aje-btn sc-fUnNpA iPhqkU aje-btn--dropdown is-medium"
      id="react-aria-:r2:"
      type="button"
    >
      Open up
      <i
        aria-hidden="true"
        class="sc-cPiJYC wZRQV aje-icon material-icons is-medium"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
</DocumentFragment>
`;

exports[`MultiSelectField > matches snapshots > when open 1`] = `
<DocumentFragment>
  <div
    class="sc-hzhKNl gQkKli aje-multiselect"
  >
    <button
      aria-controls="react-aria-:r9:"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-labelledby="react-aria-:rf: react-aria-:rb:"
      class="sc-dLNtp sc-feUYzb hkAcSC fzpzpe aje-btn sc-fUnNpA iPhqkU aje-btn--dropdown is-medium"
      data-pressed="true"
      id="react-aria-:ra:"
      type="button"
    >
      Open up
      <i
        aria-hidden="true"
        class="sc-cPiJYC wZRQV aje-icon material-icons is-medium"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
</DocumentFragment>
`;
